pipelines/oracle_grounded/schema_validation.py

```
@@ -0,0 +1,324 @@
+def _type_errors(value, schema, path):
+    """The declared type, which short-circuits every later keyword."""
+    expected = schema.get("type")
+    if expected is None:
+        return []
+    allowed = [expected] if isinstance(expected, str) else expected
+    if any(_type_matches(value, item) for item in allowed):
+        return []
+    return [f"{path} must have type {allowed!r}, got {_display_type(value)!r}"]
+
+
+def _validate(value, schema, root, path):
```

Ruff (0.16.2)

[warning] Missing return type annotation for private function `_validate`

(ANN202)